src/core/components/pin/gc-lock.js

```diff
@@ -0,0 +1,135 @@
+'use strict'
+
+const mortice = require('mortice')
+const pull = require('pull-stream')
+const EventEmitter = require('events')
+const log = require('debug')('ipfs:gc:lock')
+
+class GCLock extends EventEmitter {
+  constructor (repoOwner) {
+    super()
+
+    // Ensure that we get a different mutex for each instance of GCLock
+    // (There should only be one GCLock instance per IPFS instance, but
+    // there may be multiple IPFS instances, eg in unit tests)
+    const randId = (~~(Math.random() * 1e9)).toString(36) + Date.now()
+    this.mutex = mortice(randId, {
+      singleProcess: repoOwner
+    })
+
+    this.lockId = 0
+  }
+
+  readLock (lockedFn, cb) {
+    return this.lock('readLock', lockedFn, cb)
+  }
+
+  writeLock (lockedFn, cb) {
+    return this.lock('writeLock', lockedFn, cb)
+  }
+
+  lock (type, lockedFn, cb) {
+    if (typeof lockedFn !== 'function') {
+      throw new Error(`first argument to ${type} must be a function`)
+    }
+    if (typeof cb !== 'function') {
+      throw new Error(`second argument to ${type} must be a callback function`)
+    }
+
+    const lockId = this.lockId++
+    log(`[${lockId}] ${type} requested`)
+    this.emit(`${type} request`, lockId)
+    const locked = () => new Promise((resolve, reject) => {
+      this.emit(`${type} start`, lockId)
+      log(`[${lockId}] ${type} started`)
+      lockedFn((err, res) => {
+        this.emit(`${type} release`, lockId)
+        log(`[${lockId}] ${type} released`)
+        err ? reject(err) : resolve(res)
+      })
+    })
+
+    const lock = this.mutex[type](locked)
+    return lock.then(res => cb(null, res)).catch(cb)
```

Comment:
```suggestion
    return lock.then(res => cb(null, res), cb)
```

Is it better to do this instead? My understanding is that in this case if `cb(null, res)` throws then we get an uncaught promise rejection, whereas with `.then(res => cb(null, res)).catch(cb)` if `cb(null, res)` throws then `cb` is called _again_ with the error.

Reply:
You're right, good catch 👍